### ISSUE TITLE
on pooling connections

### ISSUE THREAD

**Issue by jackzhp** [orientechnologies/orientdb#7371]
## Expected behavior  
        ODatabaseDocumentTx db;
        if (pool == null) {
             pool = new OPartitionedDatabasePool(url_DB, user_DB, password_DB);
        }
       while(true){
        try{
            db = pool.acquire();
        }catch(OStorageException e){ //4 times
            com.orientechnologies.orient.core.exception.OStorageException: Cannot create a connection to remote server address(es): [192.168.1.9:2424]
        }catch(IllegalStateException e){ //after 4 times of OStorageException, we got:
            java.lang.IllegalStateException: You have reached maximum pool size for given partition
        }
        }

should always be:
com.orientechnologies.orient.core.exception.OStorageException: Cannot create a connection to remote server address(es): [192.168.1.9:2424]

but 

## Actual behavior  

as shown above in the code, 4 times "can not connect", then "reached maximum pool size".

and how to get notified when there is a connection available?

      while(pool.getAvailableConnections()>1){
          notify the requester.
      }

## Steps to reproduce  
as shown in the above code.

**Comment by tglman** [comment]
hi @jackzhp,

i wrote a test case following your code:
```
  @Test
  public void testFailConnection() throws IOException {
    ODatabaseDocumentTx db;
    OPartitionedDatabasePool pool = null;
    if (pool == null) {
      pool = new OPartitionedDatabasePool("remote:wrong", "wrong", "wrong");
    }
    while (true) {
      try {
        db = pool.acquire();
      } catch (OStorageException e) { //4 times
        System.out.println("one");
      } catch (IllegalStateException e) { //after 4 times of OStorageException, we got:
        e.printStackTrace();
        Assert.fail();
      }
    }
  }
```
and it ran for a while without problem, so i wasn't able to reproduce the problem

In case you reach the limit of the pool, the pool should block you and not throw an exception.

how did you setup the full test? I guess you used some multithreading it may be interesting to see that.

Regards

**Comment by tglman** [comment]
hi @jackzhp,

Where you able to connect? do you have more hint for reproduce your issue ?

Regards

**Comment by jackzhp** [comment]
To @tglman , sorry for long time, I did not respond. Originally, I was busy with other stuff, so I was not able to spend more time on this issue. Later, I decided to abandon OrientDB since it consumes too much memory. I am sorry I forgot to leave a note here.

Today saw your message again, and I read your test case. If indeed "it ran for a while without problem", then it means the pool has no size limit, you can get as many "connection" as you want. I did not try this, but I would doubt this.

By the way, 
#1. to get "can not connect", you should just disconnect the network, or do not start OrientDB service on the remote server, then the local client will tell you "can not connect", then after a few times (4), it will tell you "reached maximum pool size".

#2. previously, if the network is OK, then I can connect to DB without any problem, it just needs too much memory.

Best wishes.

**Comment by tglman** [comment]
hi @jackzhp,

Thank you for your reporting, we are redesigning the the pool in 3.0, so this issue will disappear.

Reagards

**Comment by tglman** [comment]
Hi, 

Closing this because is already fixed 3.0-RC1

Regards
